### ISSUE TITLE
scipy.io.loadmat for Matlab 7.3

### ISSUE THREAD

**Issue by skjerns** [scipy/scipy#11351]
**Is your feature request related to a problem? Please describe.**
Currently, `scipy` is unable to load MATLAB files that are saved with the new MATLAB file format.
Attempting to load them will raise an
```Python
NotImplementedError: Please use HDF reader for matlab v7.3 files
```

**Describe the solution you'd like**

I've implemented a small library that can also load MATLAB 7.3
https://github.com/skjerns/mat7.3 , using `h5py`. Do you think it makes sense to integrate this in scipy?

**Describe alternatives you've considered**

Just leaving it as it is might be actually okay, as adding it would introduce a dependency on `h5py` (which is not a dependency of scipy afaik?).

**Comment by matthew-brett** [comment]
I don't think we should add the h5py dependency unconditionally, but it could work as an optional dependency, so it gives an error message when h5py is not installed, suggesting the user install it.

The output should be consistent with the current output for other `.mat` files, unless there's a good reason for that not to be true.

**Comment by skjerns** [comment]
Alright! If this kind of optional dependency is okay as well, could I just make `mat73` a dependency (given I alter it such that it's consistent with `loadmat`)

**Comment by andyfaff** [comment]
If there was to be an optional dependency it would be direct to h5py, it would be easier to maintain.

I'm thinking that the optional dependency would be along the lines of:

```
def _function_for_loading_mat73(f):
    try:
        import h5py
    except ImportError:
        raise ImportError("Must have h5py installed to load mat73 files")

    # do the file loading
    pass
```

rather than declaring h5py in setup.py.

**Comment by skjerns** [comment]
Okay, and not import of `mat73`?

**Comment by ilayn** [comment]
I think the more verbose question is whether you would be willing to add the code to the SciPy repo directly (with the optional h5py dependency obviously).

**Comment by skjerns** [comment]
I can integrate the library into scipy (copy my code), no problem. It's just not very beautiful, so I thought it might be easier to call the function from scipy itself. On the other side, most people already have `h5py` installed (e.g. with Anaconda), but not `mat73`, so copying my code to scipy will be a better solution.

**Comment by skjerns** [comment]
@ilayn I had a look, however I have the feeling that `loadmat` itself is in a terrible state (see https://github.com/scipy/scipy/issues/7895). I think implementing a function for 7.3 that has all the confusing and outdated options of the current `loadmat` (some of which don't even make sense for 7.3 style) is not really a good idea.

**Comment by ev-br** [comment]
I think it's best to implement the reader for 7.3 under the loadmat interface using h5py as an optional dependence.  If some options from the current loadmat don't make sense for the new format, it's OK to just raise ValueErrors (or some specialized errors). 

If there are things to fix in the current loadmat code, I'm sure PRs will be welcome. This work should just be separate from the enhancement for 7.3 format.

If you feel you are willing to implement a new, better reader for legacy formats with a new interface, this can also be considered, but it's definitely more work and requires quite a bit of long-term commitment (from you, yes :-)).

**Comment by ilayn** [comment]
But we can of course help. Due to our limited attention span and time, modules get an uneven care from the maintainers so if you have a better idea we can work it out how to make it with minimum backwards incompatibility. Since people mostly moved on to newer versions of matlab I suspect it would be also welcomed by the community.

**Comment by skjerns** [comment]
I guess the main improvements right now would be to load `struct` to `dict` and `cell` to `list` instead of loading them into `ndarray.object`.

See also http://blog.nephics.com/2019/08/28/better-loadmat-for-scipy/

**Comment by pv** [comment]
Retaining backward compatibility is a constraint in additions, though. Adding new loading options, and changing behavior in previously undefined cases is generally easy, otherwise painful.

**Comment by robert-wilcox** [comment]
Is this still an issue?  I have a mat file I load that contains a SimulationOutput and I don't receive any errors.  The mat file is created by 2023b.

**Comment by skjerns** [comment]
> Is this still an issue? I have a mat file I load that contains a SimulationOutput and I don't receive any errors. The mat file is created by 2023b.

yes, because the `save` default of MATLAB2023b is still the old format, you need to explicitly name -v7.3 to save with the new format: https://de.mathworks.com/help/matlab/ref/save.html

`save("example.mat","A","B","-v7.3")`
